File: iosApp/iosApp/Modules/Confirmation/ConfirmationPresenter.swift
// Created by web3d4v on 20/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

enum ConfirmationPresenterEvent {
    case confirm
    case txSuccessCTATapped
    case txSuccessCTASecondaryTapped
    case txFailedCTATapped
    case txFailedCTASecondaryTapped
    case dismiss
}

protocol ConfirmationPresenter {
    var contextType: ConfirmationWireframeContext.`Type` { get }
    func present()
    func handle(_ event: ConfirmationPresenterEvent)
}

final class DefaultConfirmationPresenter {
    private weak var view: ConfirmationView?
    private let wireframe: ConfirmationWireframe
    private let interactor: ConfirmationInteractor
    private let context: ConfirmationWireframeContext
    
    private var txHash: String?
    private var error: Error?
    
    init(
        view: ConfirmationView,
        wireframe: ConfirmationWireframe,
        interactor: ConfirmationInteractor,
        context: ConfirmationWireframeContext
    ) {
        self.view = view
        self.wireframe = wireframe
        self.interactor = interactor
        self.context = context
    }
}

extension DefaultConfirmationPresenter: ConfirmationPresenter {
    var contextType: ConfirmationWireframeContext.`Type` {
        context.type
    }

    func present() {
        view?.update(with: viewModel())
    }

    func handle(_ event: ConfirmationPresenterEvent) {
        switch event {
        case .confirm:
            switch context.type {
            case .send, .sendNFT, .cultCastVote, .approveUniswap, .swap:
                wireframe.navigate(to: .authenticate(authenticateContext()))
            }
        case .txSuccessCTATapped:
            switch context.type {
            case .send:
                view?.dismiss(animated: true) { [weak self] in
                    self?.wireframe.navigate(to: .account)
                }
            case .sendNFT:
                wireframe.navigate(to: .nftsDashboard)
            case .cultCastVote:
                wireframe.navigate(to: .cultProposals)
            case .swap, .approveUniswap:
                wireframe.dismiss()
            }
        case .txSuccessCTASecondaryTapped:
            guard let txHash = txHash else { return }
            wireframe.navigate(to: .viewEtherscan(txHash: txHash))
        case .txFailedCTATapped:
            wireframe.dismiss()
        case .txFailedCTASecondaryTapped:
            guard let error = error else { return }
            wireframe.navigate(to: .report(error: error))
        case .dismiss:
            wireframe.dismiss()
        }
    }
}

private extension DefaultConfirmationPresenter {
    
    func viewModel() -> ConfirmationViewModel {
        let content: ConfirmationViewModel.Content
        switch context.type {
        case let .swap(swapData):
            content = contentViewModel(forSwap: swapData)
        case let .send(sendData):
            content = contentViewModel(forSend: sendData)
        case let .sendNFT(sendNFTData):
            content = contentViewModel(forSendNFT: sendNFTData)
        case let .cultCastVote(cultCastVoteData):
            content = contentViewModel(forCultCastVote: cultCastVoteData)
        case let .approveUniswap(approveUniswapData):
            content = contentViewModel(forApproveUniswap: approveUniswapData)
        }
        return .init(title: titleViewModel(), content: content)
    }
    
    func titleViewModel() -> String {
        Localized("confirmation.\(context.type.localizedTag).title")
    }
    
    func contentViewModel(
        forSwap data: ConfirmationWireframeContext.SwapContext
    ) -> ConfirmationViewModel.Content {
        let fiatFromValue = (
            data.amountFrom.toBigDec(
                decimals: data.currencyFrom.decimalsUInt
            ).mul(
                value: data.currencyFrom.fiatPrice.bigDec
            ).mul(value: Double(100).bigDec) // this is because we want 2 decimals on currency
        ).toBigInt().formatStringCurrency()
        let currencyFrom = ConfirmationViewModel.SwapViewModel.Currency(
            iconName: data.currencyFrom.iconName,
            symbol: data.currencyFrom.symbol,
            value: currencyFormatted(data.currencyFrom, value: data.amountFrom),
            usdValue: fiatFromValue
        )
        let fiatToValue = (
            data.amountTo.toBigDec(
                decimals: data.currencyTo.decimalsUInt
            ).mul(
                value: data.currencyTo.fiatPrice.bigDec
            ).mul(value: Double(100).bigDec) // this is because we want 2 decimals on currency
        ).toBigInt().formatStringCurrency()
        let currencyTo = ConfirmationViewModel.SwapViewModel.Currency(
            iconName: data.currencyTo.iconName,
            symbol: data.currencyTo.symbol,
            value: currencyFormatted(data.currencyTo, value: data.amountTo),
            usdValue: fiatToValue
        )
        let provider = ConfirmationViewModel.SwapViewModel.Provider(
            iconName: data.provider.iconName,
            name: data.provider.name,
            slippage: data.provider.slippage
        )
        // TODO: @Annon to show price here
        let feeValueInToken = "value token"
        let feeValueInUSD = "🤷🏻‍♂️"
        let estimatedFee = ConfirmationViewModel.SwapViewModel.Fee(
            value: feeValueInToken,
            usdValue: feeValueInUSD
        )
        return .swap(
            .init(
                currencyFrom: currencyFrom,
                currencyTo: currencyTo,
                provider: provider,
                estimatedFee: estimatedFee
            )
        )
    }
    
    func contentViewModel(
        forSend data: ConfirmationWireframeContext.SendContext
    ) -> ConfirmationViewModel.Content {
        let fiatValue = (
            data.amount.toBigDec(
                decimals: data.currency.decimalsUInt
            ).mul(
                value: data.currency.fiatPrice.bigDec
            ).mul(value: Double(100).bigDec) // this is because we want 2 decimals on currency
        ).toBigInt().formatStringCurrency()
        let currency = ConfirmationViewModel.SendViewModel.Currency(
            iconName: data.currency.iconName,
            symbol: data.currency.symbol,
            value: currencyFormatted(data.currency, value: data.amount),
            usdValue: fiatValue
        )
        let destination = ConfirmationViewModel.SendViewModel.Destination(
            from: Formatters.Companion.shared.networkAddress.format(
                address: data.addressFrom, digits: 8, network: data.network
            ),
            to: Formatters.Companion.shared.networkAddress.format(
                address: data.addressTo, digits: 8, network: data.network
            )
        )
        // TODO: @Annon to show price here
        let feeValueInToken = "value token"
        let feeValueInUSD = "🤷🏻‍♂️"
        let estimatedFee = ConfirmationViewModel.SendViewModel.Fee(
            value: feeValueInToken,
            usdValue: feeValueInUSD
        )
        return .send(
            .init(
                currency: currency,
                destination: destination,
                estimatedFee: estimatedFee
            )
        )
    }
    
    func contentViewModel(
        forSendNFT data: ConfirmationWireframeContext.SendNFTContext
    ) -> ConfirmationViewModel.Content {
        // TODO: Fix when supporting sending NFTs on different networks...
        let network = Network.Companion().ethereum()
        let destination = ConfirmationViewModel.SendNFTViewModel.Destination(
            from: Formatters.Companion.shared.networkAddress.format(
                address: data.addressFrom, digits: 8, network: network
            ),
            to: Formatters.Companion.shared.networkAddress.format(
                address: data.addressTo, digits: 8, network: network
            )
        )
        // TODO: @Annon to show price here
        let feeValueInToken = "value token"
        let feeValueInUSD = "🤷🏻‍♂️"
        let estimatedFee = ConfirmationViewModel.SendNFTViewModel.Fee(
            value: feeValueInToken,
            usdValue: feeValueInUSD
        )
        return .sendNFT(
            .init(
                nftItem: data.nftItem,
                destination: destination,
                estimatedFee: estimatedFee
            )
        )
    }
    
    func contentViewModel(
        forCultCastVote data: ConfirmationWireframeContext.CultCastVoteContext
    ) -> ConfirmationViewModel.Content {
        .cultCastVote(
            .init(
                action: data.approve ? Localized("approve") : Localized("reject"),
                name: data.cultProposal.title
            )
        )
    }
    
    func contentViewModel(
        forApproveUniswap data: ConfirmationWireframeContext.ApproveUniswapContext
    ) -> ConfirmationViewModel.Content {
        // TODO: @Annon to show price here
        let feeValueInToken = "value token"
        let feeValueInUSD = "🤷🏻‍♂️"
        let estimatedFee = ConfirmationViewModel.ApproveUniswapViewModel.Fee(
            value: feeValueInToken,
            usdValue: feeValueInUSD
        )
        return .approveUniswap(
            .init(
                iconName: data.currency.iconName,
                symbol: data.currency.symbol,
                fee: estimatedFee
            )
        )
    }
}

private extension DefaultConfirmationPresenter {
    
    func showTransactionInProgress() {
        let viewModel = ConfirmationViewModel(
            title: titleViewModel(),
            content: .inProgress(
                .init(
                    title: Localized("confirmation.tx.inProgress.\(context.type.localizedTag).title"),
                    message: Localized("confirmation.tx.inProgress.\(context.type.localizedTag).message")
                )
            )
        )
        view?.update(with: viewModel)
    }
}

private extension DefaultConfirmationPresenter {
    
    func showTransactionSuccess(_ response: TransactionResponse) {
        self.txHash = response.hash
        let viewModel = ConfirmationViewModel(
            title: titleViewModel(),
            content: .success(
                .init(
                    title: Localized("confirmation.tx.success.\(context.type.localizedTag).title"),
                    message: Localized("confirmation.tx.success.\(context.type.localizedTag).message"),
                    cta: Localized("confirmation.tx.success.\(context.type.localizedTag).cta"),
                    ctaSecondary: Localized("confirmation.tx.success.viewEtherScan")
                )
            )
        )
        view?.update(with: viewModel)
    }
}

private extension DefaultConfirmationPresenter {
    
    func showTransactionFailed(_ error: Error) {
        self.error = error
        let viewModel = ConfirmationViewModel(
            title: titleViewModel(),
            content: .failed(
                .init(
                    title: Localized("confirmation.tx.failed.\(context.type.localizedTag).title"),
                    error: error.localizedDescription,
                    cta: Localized("confirmation.tx.failed.\(context.type.localizedTag).cta"),
                    ctaSecondary: Localized("confirmation.tx.failed.report")
                )
            )
        )
        view?.update(with: viewModel)
    }
}

private extension DefaultConfirmationPresenter {
    
    func authenticateContext() -> AuthenticateWireframeContext {
        .init(
            title: Localized("authenticate.title.\(context.type.localizedTag)"),
            keyStoreItem: nil,
            handler: onAuthenticatedHandler()
        )
    }
    
    func onAuthenticatedHandler() -> (AuthenticateData?, KotlinError?) -> Void {
        {
            [weak self] data, error in
            if let data = data {
                self?.handleSuccessfulAuthentication(with: data.password, and: data.salt)
            } else {
                print(error ?? "No error but no data either")
            }
        }
    }
    
    func handleSuccessfulAuthentication(
        with password: String,
        and salt: String
    ) {
        showTransactionInProgress()
        broadcastTransaction(with: password, and: salt)
    }

    func broadcastTransaction(
        with password: String,
        and salt: String
    ) {
        txHash = nil
        error = nil
        switch context.type {
        case let .approveUniswap(context):
            context.onApproved((password: password, salt: salt))
            wireframe.dismiss()
        case let .swap(context):
            interactor.executeSwap(
                network: context.network,
                password: password,
                salt: salt,
                swapService: context.swapService,
                handler: { [weak self] result in
                    switch result {
                    case let .success(response):
                        self?.txHash = response.hash
                        self?.showTransactionSuccess(response)
                    case let .failure(error):
                        print("[ERROR]", error)
                        self?.showTransactionFailed(error)
                    }
                }
            )
        case let .send(data):
            interactor.send(
                addressTo: data.addressTo,
                network: data.network,
                currency: data.currency,
                amount: data.amount,
                networkFee: data.networkFee,
                password: password,
                salt: salt,
                handler: { [weak self] result in
                    switch result {
                    case let .success(response):
                        self?.showTransactionSuccess(response)
                    case let .failure(error):
                        self?.showTransactionFailed(error)
                    }
                }
            )
        case let .sendNFT(data):
            interactor.sendNFT(
                addressFrom: data.addressFrom,
                addressTo: data.addressTo,
                network: data.network,
                nft: data.nftItem,
                password: password,
                salt: salt,
                handler: { [weak self] result in
                    switch result {
                    case let .success(response):
                        self?.showTransactionSuccess(response)
                    case let .failure(error):
                        self?.showTransactionFailed(error)
                    }
                }
            )
        case let .cultCastVote(data):
            interactor.castVote(
                proposalId: data.cultProposal.id,
                support: data.approve,
                password: password,
                salt: salt,
                handler: { [weak self] result in
                    switch result {
                    case let .success(response):
                        self?.showTransactionSuccess(response)
                    case let .failure(error):
                        self?.showTransactionFailed(error)
                    }
                }
            )
        }
    }
}

private extension ConfirmationPresenter {
    
    func currencyFormatted(_ currency: Currency, value: BigInt) -> String {
        var currencyFormatted = value.formatString(
            type: .long(minDecimals: 10),
            decimals: currency.decimalsUInt
        )
        if currencyFormatted.nonDecimals.count > 10 {
            currencyFormatted = value.formatString(
                type: .long(minDecimals: 4),
                decimals: currency.decimalsUInt
            )
        } else if currencyFormatted.nonDecimals.count > 6 {
            currencyFormatted = value.formatString(
                type: .long(minDecimals: 5),
                decimals: currency.decimalsUInt
            )
        } else if currencyFormatted.nonDecimals.count > 3 {
            currencyFormatted = value.formatString(
                type: .long(minDecimals: 7),
                decimals: currency.decimalsUInt
            )
        }
        return currencyFormatted + " \(currency.symbol)"
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/confirmation/ConfirmationPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.confirmation

import com.sonsofcrypto.web3lib.formatters.Formater
import com.sonsofcrypto.web3lib.formatters.Formater.Style.Custom
import com.sonsofcrypto.web3lib.provider.model.TransactionResponse
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.legacy.NetworkFee
import com.sonsofcrypto.web3lib.types.bignum.BigDec
import com.sonsofcrypto.web3lib.types.bignum.BigInt
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3lib.utils.bgDispatcher
import com.sonsofcrypto.web3lib.utils.withUICxt
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.extensions.timeString
import com.sonsofcrypto.web3walletcore.modules.authenticate.AuthenticateData
import com.sonsofcrypto.web3walletcore.modules.authenticate.AuthenticateWireframeContext
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationPresenterEvent.Confirm
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationPresenterEvent.TxFailedCTASecondaryTapped
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationPresenterEvent.TxFailedCTATapped
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationPresenterEvent.TxSuccessCTASecondaryTapped
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationPresenterEvent.TxSuccessCTATapped
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.ApproveUniswap
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.ApproveUniswapContext
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.CultCastVote
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.CultCastVoteContext
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.Send
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.SendContext
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.SendNFT
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.SendNFTContext
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.Swap
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.SwapContext
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeContext.SwapContext.Provider
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeDestination.Account
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeDestination.Authenticate
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeDestination.CultProposals
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeDestination.Dismiss
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeDestination.NftsDashboard
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeDestination.Report
import com.sonsofcrypto.web3walletcore.modules.confirmation.ConfirmationWireframeDestination.ViewEtherscan
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

sealed class ConfirmationPresenterEvent {
    object Confirm: ConfirmationPresenterEvent()
    object TxSuccessCTATapped: ConfirmationPresenterEvent()
    object TxSuccessCTASecondaryTapped: ConfirmationPresenterEvent()
    object TxFailedCTATapped: ConfirmationPresenterEvent()
    object TxFailedCTASecondaryTapped: ConfirmationPresenterEvent()
    object Dismiss: ConfirmationPresenterEvent()
}

interface ConfirmationPresenter {
    fun present()
    fun handle(event: ConfirmationPresenterEvent)
    fun context(): ConfirmationWireframeContext
}

class DefaultConfirmationPresenter(
    private val view: WeakRef<ConfirmationView>?,
    private val wireframe: ConfirmationWireframe,
    private val interactor: ConfirmationInteractor,
    private val context: ConfirmationWireframeContext,
): ConfirmationPresenter {
    private var txHash: String? = null
    private var error: Throwable? = null
    private val scope = CoroutineScope(bgDispatcher)

    override fun present() { updateView(contentViewModel(context)) }

    override fun handle(event: ConfirmationPresenterEvent) {
        when (event) {
            is Confirm -> wireframe.navigate(Authenticate(authenticateContext()))
            is TxSuccessCTATapped -> {
                when (context) {
                    is Send -> wireframe.navigate(Dismiss { wireframe.navigate(Account) })
                    is SendNFT -> wireframe.navigate(NftsDashboard)
                    is CultCastVote -> wireframe.navigate(CultProposals)
                    is Swap -> wireframe.navigate(Dismiss())
                    is ApproveUniswap -> wireframe.navigate(Dismiss())
                }
            }
            is TxSuccessCTASecondaryTapped -> {
                val txHash = txHash ?: return
                wireframe.navigate(ViewEtherscan(txHash))
            }
            is TxFailedCTATapped -> wireframe.navigate(Dismiss())
            is TxFailedCTASecondaryTapped -> {
                val error = error ?: return
                wireframe.navigate(
                    Report(error.message ?: Localized("confirmation.tx.failed.generic.error"))
                )
            }
            is ConfirmationPresenterEvent.Dismiss -> wireframe.navigate(Dismiss())
        }
    }

    override fun context(): ConfirmationWireframeContext = context

    private fun updateView(content: ConfirmationViewModel.Content) {
        view?.get()?.update(viewModel(content))
    }

    private fun viewModel(content: ConfirmationViewModel.Content): ConfirmationViewModel =
        ConfirmationViewModel(Localized("confirmation.${context.localized}.title"), content)

    private fun inProgressViewModel(): ConfirmationViewModel.Content {
        val data = ConfirmationTxInProgressViewModel(
            Localized("confirmation.tx.inProgress.${context.localized}.title"),
            Localized("confirmation.tx.inProgress.${context.localized}.message"),
        )
        return ConfirmationViewModel.Content.TxInProgress(data)
    }

    private fun successViewModel(): ConfirmationViewModel.Content {
        val data = ConfirmationTxSuccessViewModel(
            Localized("confirmation.tx.success.${context.localized}.title"),
            Localized("confirmation.tx.success.${context.localized}.message"),
            Localized("confirmation.tx.success.${context.localized}.cta"),
            Localized("confirmation.tx.success.viewEtherScan"),
        )
        return ConfirmationViewModel.Content.TxSuccess(data)
    }

    private fun failedViewModel(error: Throwable): ConfirmationViewModel.Content {
        val data = ConfirmationTxFailedViewModel(
            Localized("confirmation.tx.failed.${context.localized}.title"),
            Localized(error.message ?: Localized("confirmation.tx.failed.generic.error")),
            Localized("confirmation.tx.failed.${context.localized}.cta"),
            Localized("confirmation.tx.failed.report"),
        )
        return ConfirmationViewModel.Content.TxFailed(data)
    }

    private fun contentViewModel(
        context: ConfirmationWireframeContext
    ): ConfirmationViewModel.Content = when (context) {
        is Swap -> contentViewModel(context.data)
        is Send -> contentViewModel(context.data)
        is SendNFT -> contentViewModel(context.data)
        is CultCastVote -> contentViewModel(context.data)
        is ApproveUniswap -> contentViewModel(context.data)
    }

    private fun contentViewModel(input: SwapContext): ConfirmationViewModel.Content {
        val data = ConfirmationSwapViewModel(
            currencyViewModel(input.currencyFrom, input.amountFrom),
            currencyViewModel(input.currencyTo, input.amountTo),
            providerViewModel(input.provider),
            networkFeeViewModel(input.networkFee),
        )
        return ConfirmationViewModel.Content.Swap(data)
    }

    private fun contentViewModel(input: SendContext): ConfirmationViewModel.Content {
        val data = ConfirmationSendViewModel(
            currencyViewModel(input.currency, input.amount),
            addressViewModel(input.addressFrom, input.addressTo, input.network),
            networkFeeViewModel(input.networkFee),
        )
        return ConfirmationViewModel.Content.Send(data)
    }

    private fun contentViewModel(input: SendNFTContext): ConfirmationViewModel.Content {
        val data = ConfirmationSendNFTViewModel(
            input.nftItem,
            addressViewModel(input.addressFrom, input.addressTo, input.network),
            networkFeeViewModel(input.networkFee),
        )
        return ConfirmationViewModel.Content.SendNFT(data)
    }

    private fun contentViewModel(input: CultCastVoteContext): ConfirmationViewModel.Content {
        val data = ConfirmationCultCastVoteViewModel(
            if (input.approve) Localized("approve") else Localized("reject"),
            input.cultProposal.title,
            networkFeeViewModel(input.networkFee),
        )
        return ConfirmationViewModel.Content.CultCastVote(data)
    }

    private fun contentViewModel(input: ApproveUniswapContext): ConfirmationViewModel.Content {
        val data = ConfirmationApproveUniswapViewModel(
            input.currency.iconName,
            input.currency.symbol,
            networkFeeViewModel(input.networkFee),
        )
        return ConfirmationViewModel.Content.ApproveUniswap(data)
    }

    private fun currencyViewModel(c: Currency, a: BigInt): ConfirmationCurrencyViewModel =
        ConfirmationCurrencyViewModel(
            c.iconName,
            c.symbol,
            Formater.currency.format(a, c, Custom(15u)),
            fiatPrice(a, c, Custom(10u), "usd")
        )

    private fun addressViewModel(
        from: String, to: String, network: Network
    ): ConfirmationAddressViewModel =
        ConfirmationAddressViewModel(
            Formater.address.format(from, 8, network),
            Formater.address.format(to, 8, network),
        )

    private fun providerViewModel(provider: Provider): ConfirmationProviderViewModel =
        ConfirmationProviderViewModel(provider.iconName, provider.name, provider.slippage)

    private fun networkFeeViewModel(fee: NetworkFee): ConfirmationNetworkFeeViewModel {
        val output = Formater.currency.format(fee.amount, fee.currency, Custom(10u))
            .toMutableList()
        val fiatOutput = fiatPrice(fee.amount, fee.currency, Custom(8u))
        output.add(Formater.Output.Normal(" ~ "))
        output.addAll(fiatOutput)
        return ConfirmationNetworkFeeViewModel(
            Localized("networkFeeView.estimatedFee"),
            Formater.currency.format(fee.amount, fee.currency, Custom(10u)),
            fee.timeString,
            fiatPrice(fee.amount, fee.currency, Custom(8u)),
        )
    }

    private fun fiatPrice(
        amount: BigInt, currency: Currency, style: Formater.Style, currencyCode: String = "usd"
    ): List<Formater.Output> {
        val value = Formater.crypto(amount, currency.decimals(), interactor.fiatPrice(currency))
        return Formater.fiat.format(
            BigDec.from(BigDec.from(value).mul(BigDec.from(100)).toBigInt()).div(BigDec.from(100)),
            style,
            currencyCode
        )
    }

    private fun authenticateContext(): AuthenticateWireframeContext =
        AuthenticateWireframeContext(
            Localized("authenticate.title.${context.localized}"), null, authenticatedHandler()
        )

    private fun authenticatedHandler(): (data: AuthenticateData?, error: Error?) -> Unit {
        return { data, error ->
            if (data != null) {
                updateView(inProgressViewModel())
                broadcastTransaction(data)
            } else {
                println(error ?: "No error but no data either")
            }
        }
    }

    private fun broadcastTransaction(data: AuthenticateData) {
        txHash = null
        error = null
        when (context) {
            is Send -> {
                scope.launch {
                    try {
                        val result = interactor.send(context.data, data.password, data.salt)
                        withUICxt { showSuccess(result) }
                    } catch (e: Throwable) {
                        withUICxt { error = e; updateView(failedViewModel(error!!)) }
                    }
                }
            }
            is Swap -> {
                scope.launch {
                    try {
                        val result = interactor.swap(
                            context.data.network, data.password, data.salt, context.data.swapService
                        )
                        withUICxt { showSuccess(result) }
                    } catch (e: Throwable) {
                        withUICxt { error = e; updateView(failedViewModel(error!!)) }
                    }
                }
            }
            is SendNFT -> {
                scope.launch {
                    try {
                        val result = interactor.sendNFT(context.data, data.password, data.salt)
                        withUICxt {
                            interactor.trackNFTAsSent(context.data.nftItem)
                            showSuccess(result)
                        }
                    } catch (e: Throwable) {
                        withUICxt { error = e; updateView(failedViewModel(error!!)) }
                    }
                }
            }
            is CultCastVote -> {
                scope.launch {
                    try {
                        val result = interactor.castVote(context.data, data.password, data.salt)
                        withUICxt { showSuccess(result) }
                    } catch (e: Throwable) {
                        withUICxt { error = e; updateView(failedViewModel(error!!)) }
                    }
                }
            }
            is ApproveUniswap -> {
                context.data.onApproved(data.password, data.salt)
                wireframe.navigate(Dismiss())
            }
        }
    }

    fun showSuccess(tx: TransactionResponse) { txHash = tx.hash; updateView(successViewModel()) }
}

private val ConfirmationWireframeContext.localized: String get() = when (this) {
    is Swap -> "swap"
    is Send -> "send"
    is SendNFT -> "sendNFT"
    is CultCastVote -> "cultCastVote"
    is ApproveUniswap -> "approveUniswap"
}
